### PR TITLE
Adds WinDivert support to Windows builds

### PR DESCRIPTION
**Pull request by smithjacobj**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: https://redmine.openinfosecfoundation.org/issues/2455

Enables IPS functionality on Windows using the open-source (LGPLv3/GPLv2)
WinDivert driver and API.

From https://www.reqrypt.org/windivert-doc.html : "WinDivert is a
user-mode capture/sniffing/modification/blocking/re-injection package
for Windows Vista, Windows Server 2008, Windows 7, and Windows 8.
WinDivert can be used to implement user-mode packet filters, packet
sniffers, firewalls, NAT, VPNs, tunneling applications, etc., without
the need to write kernel-mode code."

- adds `--windivert [filter string]` and `--windivert-forward [filter
    string]` command-line options to enable WinDivert IPS mode.
    `--windivert[-forward] true` will open a filter for all traffic. See
    https://www.reqrypt.org/windivert-doc.html#filter_language for more
    information.

Limitation: currently limited to `autofp` runmode.

Additionally:
- `tmm_modules` now zeroed during `RegisterAllModules`
- fixed Windows Vista+ `inet_ntop` call in `PrintInet`
- fixed `GetRandom` bug (nonexistent keys) on fresh Windows installs

---

This is a replacement for #3337

### CODE REVIEW

File: doc/userguide/setting-up-ipsinline-for-windows.rst [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+command in an elevated command prompt or terminal:
+
+::
+  
+  suricata -c suricata.yaml --windivert [filter string]
+
+For information on the WinDivert filter language, see
+https://www.reqrypt.org/windivert-doc.html#filter_language
+
+If Suricata is running on a gateway and is meant to protect the computers behind
```
Comment (norg) [inline doc/userguide/setting-up-ipsinline-for-windows.rst]: just minor suggestion, "protect the network" instead of "computers" :)

File: src/source-windivert.c [hunk 2]
```diff
@@ -0,0 +1,891 @@
...
+     * for direct data alloc size, and this is meaningless if large segments are
+     * coalesced before they reach WinDivert */
+    bool success = false;
+    if (wd_tv->offload_enabled) {
+        /* allocate external, if not already */
+        PacketCallocExtPkt(p, MAX_PAYLOAD_SIZE);
+
+        success = WinDivertRecv(wd_tv->filter_handle, p->ext_pkt,
+                                MAX_PAYLOAD_SIZE, &p->windivert_v.addr,
+                                &p->pktlen);
```
Comment (victorjulien) [inline src/source-windivert.c]: our coccinelle check for pkt access triggers on this:
```
Using concurrency level 28
Testing cocci file: access-pkt-packet.cocci
Invalid Packet->pktlen usage, GET_PKT_LEN macro must be used at /builds/inliniac/suricata-ci/suricata/qa/coccinelle/../../src/source-windivert.c:342
ERROR: Job failed: exit code 1
```

Comment (victorjulien) [reply src/source-windivert.c]: I guess the way to work around this would be to use a local var to get the pktlen from WinDivertRecv and then use the SET_PKT_LEN macro.

File: src/source-windivert.c [hunk 3]
```diff
@@ -0,0 +1,891 @@
...
+
+    g_wd_num++;
+
+unlock:
+    SCMutexUnlock(&g_wd_init_lock);
+
+    if (ret == 0) {
+        // stringify queue index to use as thread name descriptor
+        char wd_num_str[6];
+        wd_num_str[sizeof(wd_num_str) - 1] = 0;
```
Comment (victorjulien) [inline src/source-windivert.c]: instead of this line you can simply init the var at declaration:
```c
char wd_num_str[6] = "";
```

File: src/source-windivert.c [hunk 4]
```diff
@@ -0,0 +1,891 @@
...
+    }
+
+    for (IP_ADAPTER_ADDRESSES *if_info = if_info_list; if_info != NULL;
+         if_info = if_info->Next) {
+
+        if (WinDivertIfaceMatchFilter(wd_qv->filter_str, if_info->IfIndex)) {
+            SCLogInfo("Found adapter %s matching WinDivert filter %s",
+                      if_info->AdapterName, wd_qv->filter_str);
+
+            LiveDevice *new_ldev = malloc(sizeof(LiveDevice));
```
Comment (victorjulien) [inline src/source-windivert.c]: please use SCMalloc and check the result

Comment (victorjulien) [reply src/source-windivert.c]: actually I'd prefer SCCalloc so the struct is zeroed

File: src/source-windivert.c [hunk 5]
```diff
@@ -0,0 +1,891 @@
...
+            new_ldev->dev = SCStrdup(if_info->AdapterName);
```
Comment (victorjulien) [inline src/source-windivert.c]: check result of SCStrdup

File: src/source-windivert.c [hunk 6]
```diff
@@ -0,0 +1,891 @@
...
+ * \brief WinDivert verdict module packet entry function
+ */
+TmEcode VerdictWinDivert(ThreadVars *tv, Packet *p, void *data, PacketQueue *pq,
+                         PacketQueue *postpq)
+{
+    SCEnter();
+
+    TmEcode ret = TM_ECODE_OK;
+
+    /* \todo do we need to specifically handle tunnel packets like NFQ? */
```
Comment (victorjulien) [inline src/source-windivert.c]: comment obsolete?

File: src/source-windivert.c [hunk 7]
```diff
@@ -0,0 +1,891 @@
...
+    size_t count = (sizeof(tests) / sizeof(tests[0]));
+
+    for (size_t i = 0; i < count; i++) {
+        struct testdata test = tests[i];
+
+        bool actual = WinDivertIfaceMatchFilter(test.filter, test.if_index);
+        if (actual != test.expected) {
+            printf("WinDivertIfaceMatchFilter(\"%s\", %d) == %d, expected %d\n",
+                   test.filter, test.if_index, actual, test.expected);
+            // FAIL;
```
Comment (victorjulien) [inline src/source-windivert.c]: todo?

File: src/suricata.c [hunk 8]
```diff
@@ -2405,12 +2463,30 @@ static int ConfigGetCaptureValue(SCInstance *suri)
...
+                    break;
+                }
+
+                g_default_mtu = DEFAULT_MTU;
+                default_packet_size = DEFAULT_PACKET_SIZE;
+                break;
+#endif /* WINDIVERT */
             case RUNMODE_PCAP_DEV:
             case RUNMODE_AFP_DEV:
+            /* \bug: are PCAP/AFP supposed to fall through to strip_trailing_plus? */
```
Comment (victorjulien) [inline src/suricata.c]: can you remove this comment and add it as a ticket or a PR?

File: src/util-ioctl.c [hunk 9]
```diff
@@ -53,6 +57,8 @@
  */
 static int GetIfaceMaxHWHeaderLength(const char *pcap_dev)
 {
+    /* \bug: this isn't necessarily universal (my Ubuntu WiFi device is wlp1s0),
```
Comment (victorjulien) [inline src/util-ioctl.c]: same here: please record things like this elsewhere

File: src/util-print.c [hunk 10]
```diff
@@ -268,7 +268,16 @@ const char *PrintInet(int af, const void *src, char *dst, socklen_t size)
 {
     switch (af) {
         case AF_INET:
+#if defined(OS_WIN32) && NTDDI_VERSION >= NTDDI_VISTA
+            // because Windows has to provide a non-conformant inet_ntop, of
+            // course!
+            ; // case has to start with a statement
```
Comment (victorjulien) [inline src/util-print.c]: I prefer adding braces instead, something like:
```c
case AF_INET:
#if defined(OS_WIN32) && NTDDI_VERSION >= NTDDI_VISTA
{
  struct in_addr _src;
  memcpy(&_src, src, sizeof(struct in_addr));
  return inet_ntop(af, &_src, dst, size);
}
```

File: src/util-random.c [hunk 11]
```diff
@@ -24,8 +24,8 @@
  * SEI CERT C Coding Standard MSC30-C
  */
 
-#include "suricata-common.h"
 #include "util-random.h"
+#include "suricata-common.h"
```
Comment (victorjulien) [inline src/util-random.c]: why is this needed?

File: src/win32-misc.h [hunk 12]
```diff
@@ -24,6 +24,8 @@
 #ifndef __WIN32_MISC_H__
 #define __WIN32_MISC_H__
 
+#include <sdkddkver.h>
```
Comment (victorjulien) [inline src/win32-misc.h]: I'm not familiar with this header, is it available on all windows platforms?

Comment (victorjulien) [reply src/win32-misc.h]: it should probably be checked for existence in configure.ac

File: src/win32-syscall.c [hunk 13]
```diff
@@ -0,0 +1,1651 @@
...
+{
+    DWORD err = NO_ERROR;
+    IP_ADAPTER_ADDRESSES *if_info_list;
+
+    ULONG size = 0;
+    err = GetAdaptersAddresses(AF_UNSPEC, 0, NULL, NULL, &size);
+    if (err != ERROR_BUFFER_OVERFLOW) {
+        return err;
+    }
+    if_info_list = malloc((size_t)size);
```
Comment (victorjulien) [inline src/win32-syscall.c]: SCMalloc

File: src/win32-syscall.c [hunk 14]
```diff
@@ -0,0 +1,1651 @@
...
+    if (if_info_list == NULL) {
+        return ERROR_NOT_ENOUGH_MEMORY;
+    }
+    err = GetAdaptersAddresses(AF_UNSPEC, 0, NULL, if_info_list, &size);
+    if (err != NO_ERROR) {
+        free(if_info_list);
```
Comment (victorjulien) [inline src/win32-syscall.c]: SCFree

File: src/win32-syscall.c [hunk 15]
```diff
@@ -0,0 +1,1651 @@
...
+    }
+
+    if (*pif_info == NULL) {
+        ret = ERROR_NOT_FOUND;
+    }
+
+    return ret;
+}
+
+#if NTDDI_VERSION < NTDDI_VISTA
```
Comment (victorjulien) [inline src/win32-syscall.c]: does this still make sense? Is there something older than Vista we'd want to support? I'm thinking there is not. As this is new functionality we can easily set a minimum supported version to be something recent. (e.g. just Win10 + latest Server?)

File: src/win32-syscall.c [hunk 16]
```diff
@@ -0,0 +1,1651 @@
...
+    /* now search for the right adapter in the list */
+    IP_ADAPTER_ADDRESSES *if_info = NULL;
+    for (if_info = if_info_list; if_info != NULL; if_info = if_info->Next) {
+        /* -1 (uint) is an invalid value */
+        if (if_info->Mtu == (uint32_t)-1) {
+            continue;
+        }
+
+        /* we want to return the largest MTU value so we allocate enough */
+        if (if_info->Mtu > mtu) {
```
Comment (victorjulien) [inline src/win32-syscall.c]:
```c
mtu = MAX(mtu, if_info->Mtu);
```

**Comment by victorjulien** [comment]
Btw, I'm in the process of setting up a vm for testing.

File: src/win32-misc.h (hunk 12 above)
Comment (sgeto) [reply src/win32-misc.h]: > I'm not familiar with this header, is it available on all windows platforms?

Yes, It's a very common one.

File: src/win32-syscall.c (hunk 15 above)
Comment (smithjacobj) [reply src/win32-syscall.c]: That was a matter of assumption I made based on [this discussion](https://github.com/OISF/suricata/pull/3339#discussion_r181546625) that XP compatibility was still desirable.

File: src/source-windivert.c (hunk 3 above)
Comment (smithjacobj) [reply src/source-windivert.c]: That would zero the first index though, not the last one. Zeroing the last one means that, as long as I keep the length values correct when setting it, no c-string functions should ever overrun the buffer as the last index is always NULL.

File: src/util-random.c (hunk 11 above)
Comment (smithjacobj) [reply src/util-random.c]: It's not, not sure why I did it so I'm reverting it.

File: src/win32-misc.h (hunk 12 above)
Comment (smithjacobj) [reply src/win32-misc.h]: I'm pretty sure it's been available since Windows 2000. If we have any expectation of somebody trying to build it on Windows ME :laughing: then I guess it might need to be added to `configure.ac`

File: src/win32-syscall.c (hunk 15 above)
Comment (smithjacobj) [reply src/win32-syscall.c]: To be specific, the minimum Windows OS this functionality requires is Vista, for both the WMI APIs and WinDivert (which relies on Windows Filtering Platform introduced in Vista.)
